api/v1alpha1/bundle_types.go

```diff
@@ -51,13 +52,41 @@ type BundleSource struct {
 	Type string `json:"type"`
 	// Image is the bundle image that backs the content of this bundle.
 	Image *ImageSource `json:"image,omitempty"`
+	// Git is the git repository that backs the content of this Bundle.
+	Git *GitSource `json:"git,omitempty"`
 }
 
 type ImageSource struct {
 	// Ref contains the reference to a container image containing Bundle contents.
 	Ref string `json:"ref"`
 }
 
+type GitSource struct {
+	// Repository is a URL link to the git repository containing the bundle.
+	// Repository is required and the URL should be parsable by a standard git tool.
+	Repository string `json:"repository"`
+	// Directory refers to the location of the bundle within the git repository.
+	// Directory is optional and if not set defaults to ./manifests.
+	Directory string `json:"directory,omitempty"`
+	// Ref configures the git source to clone a specific branch, tag, or commit
+	// from the specified repo. Ref is required, and exactly one field within Ref
+	// is required. Setting more than one field or zero fields will result in an
+	// error.
+	Ref GitRef `json:"ref"`
+}
+
+type GitRef struct {
+	// Branch refers to the branch to checkout from the repository.
```

Comment:
> follow-up: validate (either in code or schema validation?) that only a single branch/tag/commit has been specified?

In a similar vein - should we also make sure that at least one is specified? As it currently sits it will pass if none of these are set - right?

Reply:
That's a good point -- I do verify that in the code here: https://github.com/operator-framework/rukpak/pull/179/files#diff-fe812db0d942803e1f21fe81bd691d1c62d0a438ba3f35309bb26cb5339970f9R66 but we would want to enforce that at the schema level as well if possible. I'm not sure how well we can mix oneOf with an at least one idea but we should try. 

We also have the Bundle webhook as a good place to drop some of this validation in as well.

Reply:
I like the idea of the Bundle webhook handling that logic.

Reply:
IMO, anything that _can_ be expressed in CRD schema should go in CRD schema. There are features coming to kubernetes that will make it easier to include logic in CRD validations via CEL expressions, so once that lands it would be nice if we could get rid of whatever's left in our validation webhooks.